src/transformers/models/fastspeech2/modeling_fastspeech2.py

```diff
@@ -0,0 +1,698 @@
+        self.ln2 = nn.LayerNorm(config.var_pred_hidden_dim)
+        self.proj = nn.Linear(config.var_pred_hidden_dim, 1)
+
+    def forward(self, x):
+        # Input: B x T x C; Output: B x T
+        x = self.conv1(x.transpose(1, 2)).transpose(1, 2)
+        x = self.dropout_module(self.ln1(x))
+        x = self.conv2(x.transpose(1, 2)).transpose(1, 2)
+        x = self.dropout_module(self.ln2(x))
+        return self.proj(x).squeeze(dim=2)
+
+
+class VarianceAdaptor(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.length_regulator = LengthRegulator()
+        self.duration_predictor = VariancePredictor(config)
+        self.pitch_predictor = VariancePredictor(config)
+        self.energy_predictor = VariancePredictor(config)
+
+        n_bins, steps = self.config.var_pred_n_bins, self.config.var_pred_n_bins - 1
+        self.pitch_bins = torch.linspace(config.pitch_min, config.pitch_max, steps)
+        self.embed_pitch = nn.Embedding(n_bins, config.encoder_embed_dim)
+        self.energy_bins = torch.linspace(config.energy_min, config.energy_max, steps)
+        self.embed_energy = nn.Embedding(n_bins, config.encoder_embed_dim)
+
+    def get_pitch_emb(self, x, tgt=None, factor=1.0):
+        out = self.pitch_predictor(x)
+        bins = self.pitch_bins.to(x.device)
+        if tgt is None:
+            out = out * factor
+            emb = self.embed_pitch(torch.bucketize(out, bins))
+        else:
+            emb = self.embed_pitch(torch.bucketize(tgt, bins))
+        return out, emb
+
+    def get_energy_emb(self, x, tgt=None, factor=1.0):
+        out = self.energy_predictor(x)
+        bins = self.energy_bins.to(x.device)
+        if tgt is None:
+            out = out * factor
+            emb = self.embed_energy(torch.bucketize(out, bins))
+        else:
+            emb = self.embed_energy(torch.bucketize(tgt, bins))
+        return out, emb
+
+    def forward(
+        self,
+        x,
+        padding_mask,
+        durations=None,
+        pitches=None,
+        energies=None,
+        d_factor=1.0,
+        p_factor=1.0,
+        e_factor=1.0,
+    ):
+        # x: B x T x C
+        log_dur_out = self.duration_predictor(x)
+        dur_out = torch.clamp(torch.round((torch.exp(log_dur_out) - 1) * d_factor).long(), min=0)
+        dur_out.masked_fill_(padding_mask, 0)
+
+        pitch_out, pitch_emb = self.get_pitch_emb(x, pitches, p_factor)
+        x = x + pitch_emb
+        energy_out, energy_emb = self.get_energy_emb(x, energies, e_factor)
+        x = x + energy_emb
+
+        x, out_lens = self.length_regulator(x, dur_out if durations is None else durations)
+
+        return x, out_lens, log_dur_out, pitch_out, energy_out
+
+
+class Postnet(nn.Module):
+    def __init__(self, in_dim, n_channels, kernel_size, n_layers, dropout):
+        super(Postnet, self).__init__()
+        self.convolutions = nn.ModuleList()
+        assert kernel_size % 2 == 1
```

Comment:
No assert in the model code base. Use a test and raise an appropriate error with a helpful message.

Reply:
I'm removing the `assert` entirely becaue `FastSpeech2Config` already raises a `ValueError` if `kernel_size` is even. Does that sound okay?